File: tournaments/prefixFunctionNaive/prefixFunctionNaive.cpp
std::vector<int> prefixFunctionNaive(std::string s) {

  std::vector<int> result;

  for (int i = 0; i < s.size(); i++) {
    result.push_back(0);
    for (result[i] = i; result[i] >= 0; result[i]--) {
      bool matches = true;
      for (int j = i - result[i] + 1; j <= i; j++) {
        if (s[j] != s[j - i + result[i] - 1]) {
          matches = false;
          break;
        }
      }
      if (matches) {
        break;
      }
    }
  }

  return result;
}

File: tournaments/prefixFunctionNaive/prefixFunctionNaive.js
function prefixFunctionNaive(s) {
    const n = s.length;
    const result = new Array(n).fill(0);
    let index = 1;
    let lastPrefixLength = 0;

    while (n > index) {
        if (s[index] === s[lastPrefixLength]) {
            result[index++] = ++lastPrefixLength;
        } else {
            if (!lastPrefixLength) {
                result[index++] = lastPrefixLength;
            } else {
                lastPrefixLength = result[lastPrefixLength - 1];
            }
        }
    }

    return result;
}
